Add APIs to reroute SDK logs on Darwin platforms

cliffamzn opened the pull request:
[Feature]

We need to collect verbose logs to assist in debugging and root causing issues that customers are experiencing. By default the Matter SDK will log with os_log which won't be captured by application loggers.

[Solution]

Added a vanilla C bridge that can be used to reroute the logs from the C++ SDK on all platforms. Added public APIs to the Darwin Framework that allows rerouting the Darwin-specific logs and the C++ SDK logs via the C bridge.

[Testing]

A local build succeeds for My Mac (Intel x86 64), Any iPhone, and the iPhone 11 simulator. Built the Framework into an iOS application and verified that there are no crashes during Matter commissioning / provisioning and that the logs are rerouted as expected.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRLogger.h:
```diff
@@ -0,0 +1,89 @@
+ *
+ * @note Messages will only be passed to this method if `isLoggingEnabledAtLevel` is `true` for the given log level.
+ */
+- (void)logAtLevel:(MTRLogLevel)level message:(NSString *)message;
```
We need to ensure that this is never called on a client-provided MTRLogger while running on the Matter dispatch queue.

That means either we need to create our own dispatch queue inside the framework for making these calls, dispatch async to it internally, and document that the API consumer cannot rely on these calls happening on any particular dispatch queue, or we need to have the framework consumer provide a dispatch queue to use when they provide an MTRLogger.

cliffamzn replied:
I changed it to use the chip worker dispatch queue.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRLogger.h:
```diff
@@ -0,0 +1,89 @@
+/**
+ * A static entry point for configuring logging in the Matter C++ and Objective C SDKs.
+ */
+@interface MTRLoggingConfiguration : NSObject {
+}
```
```suggestion
@interface MTRLoggingConfiguration : NSObject
```

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRLogger.h:
```diff
@@ -0,0 +1,89 @@
+ * If no logger has been [set](@ref setLogger), all messages at all log levels will be logged
+ * to the `os_log`.
+ *
+ * @returns The `MTRLogger` that will be used to log messages from the Matter SDK.
```
What does this return if `setLogger` never got called?

cliffamzn replied:
Clarified message

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRLogging.h:
```diff
@@ -16,12 +16,70 @@
+ * @note We declare the public API of `MTRLogging` in the `MTRLogger.h` public header file, while the non-public
+ *       module API is declared in the original `MTRLogging.h`.
+ */
+@interface MTRLogging : NSObject {
+}
```
```suggestion
@interface MTRLogging : NSObject
```

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRLogging.mm:
```diff
@@ -0,0 +1,184 @@
+    pthread_threadid_np(NULL, &ktid);
+    uint64_t pid = getpid();
+
+    // Delegate the log to the MTRLogging system
```
We need to make sure all logs from the framework/SDK go to `os_log`.  They can go somewhere else in addition to that, but they have to go to `os_log`.

cliffamzn replied:
Can you clarify why that's useful? In this case the client is identifying they want to overwrite the os logger behavior and push it to a separate log. Why do we want to force them to additionally use os_log as well?

cliffamzn replied:
Changed so that it also goes to os_log

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRLogging.mm:
```diff
@@ -0,0 +1,184 @@
+
++ (id<MTRLogger>)getLogger
+{
+    return singleton_logger;
```
This can return nil (if no logger was set), but the API definition promises to never return nil here.

cliffamzn replied:
Set a default in case it is null

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRLogging.mm:
```diff
@@ -0,0 +1,184 @@
+    // Explicitly NULL-terminate the formatted string for safety.
+    formatted_msg[sizeof(formatted_msg) - 1] = '\0';
+
+    // Collect the process and thread information to match the original CHIPLogging implementation
+    // in src/platform/Darwin/Logging.cpp
```
Can this share code with that implementation, ideally, instead of duplicating it and needing the two to stay in sync but having no indication over there that this code even exists?

cliffamzn replied:
Yes, moved some common code together

bzbarsky-apple reviewed (changes requested):
@ksperling-apple how does this interact with the logging changes you were thinking about?

ksperling-apple commented:
I'm looking at hooking the CHIP log macros more directly into `os_log_*` on Darwin so that we can take advantage of the efficient storage and built-in fields for sub-components, timestamp, tid, pid, instead of re-inventing the wheel and formatting everything into one big string first.

I was also considering putting support for log redirection behind a CHIP compile time flag (e.g. `CHIP_LOG_REDIRECTION`), and make it so redirection is in addition to rather than replaces the system log (if needed this could also be configurable by making the log redirection tri-state).

The Matter framework logging should then probably be pointed at the CHIP logging macros, rather than being a completely separate layer with it's own redirection.

cliffamzn commented on src/darwin/Framework/CHIP/MTRLogger.h:
```diff
@@ -0,0 +1,89 @@
+/**
+ * An enumeration of the log levels supported by the Matter SDK.
+ */
+typedef NS_ENUM(NSUInteger, MTRLogLevel) { MTRLogLevelFault = 0, MTRLogLevelError, MTRLogLevelInfo, MTRLogLevelDebug };
```
It's used below 

``#define MTR_LOG_FAULT(format, ...) (mtr_log(MTRLogLevelFault, MTR_C_STRING_TO_NSSTRING(format), ##__VA_ARGS__))```

bzbarsky-apple replied:
Yes, but `MTR_LOG_FAULT` is unused.

cliffamzn replied:
Removed

cliffamzn commented:
@ksperling-apple I'd love to better understand the reasoning behind having redirection an additional operation rather than a replacement one. What kind of advantages do we get from both writing to the OS log as well as the redirected log if the caller has specified they want it going elsewhere? To be clear, i'm not against it, i'm more so interested in the thought process behind it.

bzbarsky-apple commented:
> The Matter framework logging should then probably be pointed at the CHIP logging macros

Yes.  We have wanted to do that anyway for a while....

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRLogging.mm:
```diff
@@ -0,0 +1,174 @@
+{
+    if (chipWorkQueue == nil)
+    {
+        chipWorkQueue = chip::DeviceLayer::PlatformMgrImpl().GetWorkQueue();
```
No, the whole point is we do _not_ want to call out into consumer code on that work queue.

cliffamzn replied:
Got it, created a new queue altogether.

cliffamzn commented:
Will look at test failures as well in conjunction with responding to feedback

cliffamzn commented on src/lib/support/logging/CHIPLogging.cpp:
```diff
@@ -218,12 +218,13 @@ void LogV(uint8_t module, uint8_t category, const char * msg, va_list args)
-        Platform::LogV(moduleName, category, msg, args);
+        // if the redirect is going to read these args we need to make a copy
+        // of them otherwise the va_list may be modified and unreadable
+        va_list copy;
```
i'm not in love with this, would appreciate suggestions on how we might handle this better if we need to make the call to both the platform logger as well as the redirected logger

cliffamzn commented:
> I was also considering putting support for log redirection behind a CHIP compile time flag (e.g. CHIP_LOG_REDIRECTION), and make it so redirection is in addition to rather than replaces the system log (if needed this could also be configurable by making the log redirection tri-state).

@ksperling-apple How do you anticipate this compile time flag working for iOS apps that want to redirect logs and are using the built in Matter framework? Wouldn't this always be set to true?

I realize this solution doesn't quite fit super well since i'm seeing some errors where callers of the logger expect the os_log format but since the macro is assuming a CString format, it doesn't fit.

`error: using 'public' format specifier annotation`
https://github.com/project-chip/connectedhomeip/actions/runs/3343092473/jobs/5535960142#step:13:1810

Arguably the callers of the macro shouldn't be assuming that we're using an os_log format but it also looks like this macro isn't totally compatible with the chip::Logging one which takes in a module id and category. This would fit in line with your thought about migrating them all to the ChipLogging macros as well.

So if we take a step back, @ksperling-apple  / @bzbarsky-apple do you have any thoughts / suggestions on a path forward here? I'm thinking as a start, we can modify this current PR so that all callers assuming the os_log format would be standardized. We can then have another change to migrate everything to the ChipLog and get rid of the MTR_LOG macros and include module id / category to all instances that call MTR_LOG.

ksperling-apple commented:
> @ksperling-apple I'd love to better understand the reasoning behind having redirection an additional operation rather than a replacement one. What kind of advantages do we get from both writing to the OS log as well as the redirected log if the caller has specified they want it going elsewhere? To be clear, i'm not against it, i'm more so interested in the thought process behind it.
Primarily so that when a sysdiagnose is taken we also get CHIP / Matter logs even if the application chose to also capture them.

> If this is required, we can make the change here and remove the else clause so that the platform logger is always invoked: https://github.com/project-chip/connectedhomeip/blob/master/src/lib/support/logging/CHIPLogging.cpp#L219-L226
That would be fine by me (I also don't mind it being a `CHIP_*` compile time option if there's a desire to not have this happen on some platforms.)

Overall since we want to align the Matter framework logging with the CHIP logging, I think we should also align the redirection mechanisms:
* Use `Error` / `Progress` / `Detail` as the level names (suitably prefixed with e.g. `MTRLog`)
* Have the log function / method take similar arguments, i.e. module, level, format string, arguments (as va_list)

Another mismatch is that in your PR you're providing a filtering callback which CHIP doesn't support, though since the callback only gets the level (rather than e.g. module + level) we can probably just align this with the CHIP layer and just allow the level to be configured via a setter.

The last question in my mind is whether at the Darwin framework level this should be exposed as a protocol as you've done or just simple a `MTRSetLogRedirectCallback(callback, level)` C function that maps directly to the CHIP layer. Given the low level nature of this feature I'm not really sure we're gaining much by exposing it via an ObjC class + protocol.

cliffamzn commented:
@ksperling-apple - i've reworked the code to remove the MTR_LOG changes so that its strictly pulling out logs from the SDK. I also changed the level names as suggested and removed the filtering callbacks.

I think this sets us up to have a follow up PR to change MTR_LOG to move directly to the CHIP logger. Once that is done, the custom logger should pick up the new log messages naturally.

ksperling-apple commented:
I think this can be closed as the functionality is covered by #23528

cliffamzn commented:
> I think this can be closed as the functionality is covered by #23528

Agreed, will close this one out.
